logicrunner/preprocessor/templates/wrapper.go.tpl

```diff
@@ -94,40 +96,69 @@ func INSMETHOD_GetPrototype(object []byte, data []byte) ([]byte, []byte, error)
 }
 
 {{ range $method := .Methods }}
-func INSMETHOD_{{ $method.Name }}(object []byte, data []byte) ([]byte, []byte, error) {
+func INSMETHOD_{{ $method.Name }}(object []byte, data []byte) (newState []byte, result []byte, err error) {
 	ph := common.CurrentProxyCtx
 	ph.SetSystemError(nil)
+
 	self := new({{ $.ContractType }})
 
 	if len(object) == 0 {
-		return nil, nil, &foundation.Error{ S: "[ Fake{{ $method.Name }} ] ( INSMETHOD_* ) ( Generated Method ) Object is nil"}
+		err = &foundation.Error{ S: "[ Fake{{ $method.Name }} ] ( INSMETHOD_* ) ( Generated Method ) Object is nil"}
+		return
 	}
 
-	err := ph.Deserialize(object, self)
+	err = ph.Deserialize(object, self)
 	if err != nil {
-		e := &foundation.Error{ S: "[ Fake{{ $method.Name }} ] ( INSMETHOD_* ) ( Generated Method ) Can't deserialize args.Data: " + err.Error() }
-		return nil, nil, e
+		err = &foundation.Error{ S: "[ Fake{{ $method.Name }} ] ( INSMETHOD_* ) ( Generated Method ) Can't deserialize args.Data: " + err.Error() }
+		return
 	}
 
 	{{ $method.ArgumentsZeroList }}
 	err = ph.Deserialize(data, &args)
 	if err != nil {
-		e := &foundation.Error{ S: "[ Fake{{ $method.Name }} ] ( INSMETHOD_* ) ( Generated Method ) Can't deserialize args.Arguments: " + err.Error() }
-		return nil, nil, e
+		err = &foundation.Error{ S: "[ Fake{{ $method.Name }} ] ( INSMETHOD_* ) ( Generated Method ) Can't deserialize args.Arguments: " + err.Error() }
+		return
 	}
 
-{{ if $method.Results }}
-	{{ $method.Results }} := self.{{ $method.Name }}( {{ $method.Arguments }} )
-{{ else }}
-	self.{{ $method.Name }}( {{ $method.Arguments }} )
-{{ end }}
+	{{ $method.ResultDefinitions }}
+
+	serializeResults := func() error {
+		return ph.Serialize(
+			foundation.Result{Returns:[]interface{}{ {{ $method.Results }} }},
+			&result,
+		)
+	}
+
+	needRecover := true
+	defer func() {
+		if !needRecover {
+			return
+		}
+		if r := recover(); r != nil {
+			recoveredError := errors.Wrap(errors.Errorf("%v", r), "Failed to execute method (panic)")
+			recoveredError = ph.MakeErrorSerializable(recoveredError)
+
+			if PanicIsLogicalError {
+				ret{{ $method.LastErrorInRes }} = recoveredError
+
+				newState = object
+				err = serializeResults()
```

Comment:
this part doesn't look correct, it should be:
```
err = serializeResults()
if err != nil {
    newState = object
}
```

Reply:
fixed